### PR TITLE
Measurement edit and store via annotation

### PR DESCRIPTION
**Pull request by dsuren1**
## Description
This PR adds the feature of editing and storing measurement via annotation. 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**
#5941 

**What is the new behavior?**
- User can navigate to the measurement component from annotation to perform editing of measurement. Edit is restricted to the type of measurement added. 
- User can save the measurement and export the result back to annotation to be saved. (Update)

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information
Ref: https://github.com/geosolutions-it/austrocontrol-C125/issues/159

### CODE REVIEW

File: web/client/components/mapcontrols/annotations/AnnotationsEditor.jsx [hunk 1]
```diff
@@ -447,6 +451,16 @@ class AnnotationsEditor extends React.Component {
                     onUnselectFeature={this.props.onResetCoordEditor}
                 />
                 }
+                {
+                    isEditMeasure && <Button className="btn btn-primary measure" onClick={()=> {
+                        const editProperties = this.props.editing?.properties;
+                        // Excluding geometry of type Point as they are labels of measurement
+                        const features = editProperties?.type === 'Measure' && this.props.editing.features.filter(f=> f.geometry.type !== 'Point');
```
Comment (MV88) [inline web/client/components/mapcontrols/annotations/AnnotationsEditor.jsx:458]: can this type be a constant and replace all the usage of the string "Measure"?
like in utils or so put
export const MEASURE_TYPE = "Measure";

File: web/client/translations/data.it-IT.json [hunk 2]
```diff
@@ -696,6 +696,7 @@
             "newMeasure": "Nuova annotazione",
             "selectTool": "Seleziona uno strumento di misura",
             "exportToGeoJSON": "Esporta in GeoJSON",
+            "saveMeasure": "Salva la misurazione nell'annotazione",
```
Comment (MV88) [inline web/client/translations/data.it-IT.json:699]: i think is better to use plural by default, please review also other translations

```suggestion
            "saveMeasure": "Salva le misure nell'annotazione",
```

File: web/client/translations/data.en-US.json [hunk 3]
```diff
@@ -692,6 +692,7 @@
             "trueBearingLabel": "True bearing",
             "formula": "Formula for distance calculation",
             "showLabel": "Show measurement label",
+            "saveMeasure": "Save measurement to annotation",
```
Comment (MV88) [inline web/client/translations/data.en-US.json:695]:
```suggestion
            "saveMeasure": "Save measurements to annotation",
```

File: web/client/translations/data.es-ES.json [hunk 4]
```diff
@@ -1072,6 +1073,7 @@
             "zoomTo": "Zoom a la anotación",
             "zoomToGeometry": "Zoom a geometría",
             "removeGeometry": "Eliminar",
+            "editMeasurement": "Editar medida",
```
Comment (MV88) [inline web/client/translations/data.es-ES.json:1076]:
```suggestion
            "editMeasurement": "Editar Medida",
```

File: web/client/translations/data.it-IT.json [hunk 5]
```diff
@@ -1084,6 +1085,7 @@
             "cancelModalGeom": "Annulla",
             "undo": "Sei sicuro di voler abbandonare la modifica in corso?",
             "title": "Annotazioni",
+            "editMeasurement": "Modifica misurazione",
```
Comment (MV88) [inline web/client/translations/data.it-IT.json:1088]:
```suggestion
            "editMeasurement": "Modifica Misura",
```

**Review by MV88 — CHANGES_REQUESTED** [review]
@tdipisa i think that there is a missing part in the requirements that could be added or that can be added in the future. 
You cannot style differently the various measures you do, just edit geom. it this part of a future improvement?

@dsuren1  the code seems well written and covered by tests. please have a look at my comments to see if I misjudged something 

### BUGS
> The edit mode of measure will not allow to switch between area, bearing or line, it will allow to only edit the existing measurement type and save or close

- [ ] When you go to the measure tool from the annotation, what you can do is just add more measures or clear them, not edit via a selection or so. 
- [ ] When you have multiple measures the type of measure tool are enabled, but no editing is possible

- [ ] when i edit  multiple measures annotation and clear them all, and draw a new one i can see no longer the save button, but i can add as new annotation while in reality i was coming from edit mode of annotation. 

- [ ] if i continue from previous point and add as new i have duplication of annotation the old one and the new one has the same name

- [ ] when i edit an annotation via measurement and i clear measures and close the measure tool or just close the measure tool, the other annotation are not showed again. to show them again i need to select one from the annotation list and go back to the list

- [ ] when importing [this (rename as JSON)](https://github.com/geosolutions-it/MapStore2/files/5310565/measurement.annotations.with.non.measure.ann.json.TXT) exported annotations their style is highlighted. is this intended?
image

**Comment by dsuren1** [comment]
Thanks @MV88  :)

Kindly find below the answers that should help with the review 
 
> * [ ]  When you go to the measure tool from the annotation, what you can do is just add more measures or clear them, not edit via a selection or so.
> * [ ]  When you have multiple measures the type of measure tool are enabled, but no editing is possible

For this to work, we need to add this config. As this config is project specific, I didn't add this to Mapstore
`{
                "name": "Measure",
                "cfg": {
                  "defaultOptions": {
                    "showAddAsAnnotation": true,
                    "showCoordinateEditor": true
                  }
                }
            }`
> * [ ]  when i edit  multiple measures annotation and clear them all, and draw a new one i can see no longer the save button, but i can add as new annotation while in reality i was coming from edit mode of annotation.

Yes, when you clear, you can only add measure as new annotation. Else you can save the measure as annotation

> * [ ]  if i continue from previous point and add as new i have duplication of annotation the old one and the new one has the same name

It might seem to be duplicate (coz of the name, as this is generated based on the measurement type selected) but the id is new and this is a new annotation

> * [ ]  when i edit an annotation via measurement and i clear measures and close the measure tool or just close the measure tool, the other annotation are not showed again. to show them again i need to select one from the annotation list and go back to the list

I will look into this.

> * [ ]  when importing [this (rename as JSON)](https://github.com/geosolutions-it/MapStore2/files/5310565/measurement.annotations.with.non.measure.ann.json.TXT) exported annotations their style is highlighted. is this intended?
>   image

The annotation json has highlight property set to true in it

@MV88  Kindly let me know thoughts

**Comment by MV88** [comment]
> Thanks @MV88 :)
> 
> Kindly find below the clarifications that should help with the review
> 
> > * [ ]   When you go to the measure tool from the annotation, what you can do is just add more measures or clear them, not edit via a selection or so.
> > * [ ]   When you have multiple measures the type of measure tool are enabled, but no editing is possible
> 
> For this to work, we need to add this config. As this config is project specific, I didn't add this to Mapstore
> `{ "name": "Measure", "cfg": { "defaultOptions": { "showAddAsAnnotation": true, "showCoordinateEditor": true } } }`

ok, i did not knew it, but this is still weird, probably we should have that flag set to true as default? @tdipisa 

> > * [ ]   when i edit  multiple measures annotation and clear them all, and draw a new one i can see no longer the save button, but i can add as new annotation while in reality i was coming from edit mode of annotation.
> 
> Yes, when you clear, you can only add measure as new annotation. Else you can save the measure as annotation

@tdipisa is this intended ?
i think this is still not clear. when i delete all measures it should not be supposed to create a new annotation, just edit the one 
i think it matters if the measure tool is opened by itself or from annotation plugin.

> > * [ ]   if i continue from previous point and add as new i have duplication of annotation the old one and the new one has the same name
> 
> It might seem to be duplicate (coz of the name, as this is generated based on the measurement type selected) but the id is new and this is a new annotation

no no this was being duplicated, please check it out

> > * [ ]   when i edit an annotation via measurement and i clear measures and close the measure tool or just close the measure tool, the other annotation are not showed again. to show them again i need to select one from the annotation list and go back to the list
> 
> I will look into this.
 
ok

> > * [ ]   when importing [this (rename as JSON)](https://github.com/geosolutions-it/MapStore2/files/5310565/measurement.annotations.with.non.measure.ann.json.TXT) exported annotations their style is highlighted. is this intended?
> >   image
> 
> The annotation json has highlight property set to true in it

but when i export it i have no highlight selection active because the mouse in on the export all annotations button.

**Comment by tdipisa** [comment]
@dsuren1 @MV88 before doing further steps on this I would like to review the tool thogether with @dsuren1.

**Comment by tdipisa** [comment]
@MV88 this PR is ready for a new review. Thank you.

**Review by MV88 — CHANGES_REQUESTED** [review]
Hi @dsuren1 

i have found some small problems

- [ ] When i go from annotation to measure tool and then i close the measure tool the draw interaction remains active

- [ ] in the coordinate editor of measure tool the the tooltips for apply changes the tooltips remain visible even if the button is disabled
image
You can see how @vlt1 has resolved it recently by using a CustomOverlayTrigger

- [ ] there is a problem where you can end up drawing a measure and you can lose the save button 
  - open measure and draw a line
  - add as annotation and edit that line
  - deselect line tool (it remains active and cleares all the measures)
  - deselect again and reselect it or another one
  - draw something --> you see the add as new annotation --> you should seethe save button

@tdipisa as @dsuren1  has suggested me, we can remove the possibility to delesect a tool in the measure tool this will prevent this problem and if you want to clear measures you have the button or the coordinate editor

**Comment by tdipisa** [comment]
> @tdipisa as @dsuren1 has suggested me, we can remove the possibility to delesect a tool in the measure tool this will prevent this problem and if you want to clear measures you have the button or the coordinate editor

I agree with this

**Review by MV88 — APPROVED** [review]
nice job

**Pull request merged**